### PR TITLE
NativeAOT-LLVM: Init callback declaration

### PR DESCRIPTION
**Pull request by yowl**
This PR accepts the recommendation from [NativeAOT_StaticInitialization](https://github.com/dotnet/runtimelab/issues/2359#issuecomment-1651236811) and teherby removes `NativeAOT_StaticInitialization` and its initialisation.  This enables the runtime to be initialised without the use of static c++ ctors which are not invoked presently for WASI reactor components.

https://github.com/WebAssembly/wasi-libc/pull/74
https://github.com/WebAssembly/WASI/issues/13

Instead the initialisation from a WASI reactor component is done via the wasi-libc `_initialize` function.  The call to this function is done in `/Bootstrap/main.cpp` which is`#if def`ed in for WASI and `NATIVEAOT_DLL`

WASI reactor components can now be compiled and executed (although the WIT codegen is not ready yet).

This change does introduce the problem of intialisation of `CustomNativeMain` from a WASI program as it is not distnguishable from a wasi reactor compoent, and the crt1.o for the program case does not contain `_initialize`, `_intialize` is only present for the reactor version of crt1 https://github.com/WebAssembly/wasi-libc/blob/9f51a7102085ec6a6ced5778f0864c9af9f50000/libc-bottom-half/crt/crt1-reactor.c#L7-L27

To solve the problem of no `_initialize` for WASI + `NATIVEAOT_DLL` from a `CustomNativeMain` I added a `__weak` dummy function.

### CODE REVIEW

File: src/coreclr/nativeaot/Bootstrap/main.cpp [hunk 1]
```diff
@@ -169,10 +169,17 @@ extern "C" int __managed__Main(int argc, char* argv[]);
 #else
 #define NATIVEAOT_ENTRYPOINT __managed__Startup
 extern "C" void __managed__Startup();
+#ifdef HOST_WASM
+extern "C" void _initialize();
```
Comment (jkotas) [inline src/coreclr/nativeaot/Bootstrap/main.cpp:176]: Add a comment about where `_initialize` comes from and what it does?

**Comment by yowl** [comment]
Worth mentioning here is what node.js does.  

At the bottom of this document https://nodejs.org/api/wasi.html  it describes how the node.js runtime initializes WASI modules/compoents.  Ive not tested this because I have been focusing on wasmtime because it looks to be leading the way in terms of WIT support and  wasm component linking and running.  However on the surface it looks like with this change, if node.js was used as the host, then `_initialize` would be called twice resulting in failure.

File: src/coreclr/nativeaot/Bootstrap/main.cpp [hunk 2]
```diff
@@ -169,10 +168,28 @@ extern "C" int __managed__Main(int argc, char* argv[]);
 #else
 #define NATIVEAOT_ENTRYPOINT __managed__Startup
 extern "C" void __managed__Startup();
+// _initialize is a function generated by the WASI SDK libc that calls the LLVM synthesized __wasm_call_ctors function for reactor components:
+// https://github.com/WebAssembly/wasi-libc/blob/9f51a7102085ec6a6ced5778f0864c9af9f50000/libc-bottom-half/crt/crt1-reactor.c#L7-L27
+// We define and call it for NATIVEAOT_DLL and TARGET_WASI to call all the global c++ static constructors.  This ensures the runtime is intialized
```
Comment (jkotas) [inline src/coreclr/nativeaot/Bootstrap/main.cpp]:
```suggestion
// We define and call it for NATIVEAOT_DLL and TARGET_WASI to call all the global c++ static constructors.  This ensures the runtime is initialized
```

File: src/coreclr/nativeaot/Bootstrap/main.cpp [hunk 3]
```diff
@@ -169,10 +168,28 @@ extern "C" int __managed__Main(int argc, char* argv[]);
...
+// when calling into WebAssembly Component Model components
+#if defined(NATIVEAOT_DLL) && defined(TARGET_WASI)
+extern "C" void _initialize();
+
+// CustomNativeMain programs are built using the same libbootstrapperdll as NATIVEAOT_DLL but wasi-libc will not provide an _initialize implemeentation,
```
Comment (jkotas) [inline src/coreclr/nativeaot/Bootstrap/main.cpp]:
```suggestion
// CustomNativeMain programs are built using the same libbootstrapperdll as NATIVEAOT_DLL but wasi-libc will not provide an _initialize implementation,
```

File: src/coreclr/nativeaot/Bootstrap/main.cpp [hunk 4]
```diff
@@ -169,10 +168,28 @@ extern "C" int __managed__Main(int argc, char* argv[]);
...
+// so create a dummy one here and make it weak to allow wasi-libc to provide the real implmentation for WASI reactor components.
```
Comment (jkotas) [inline src/coreclr/nativeaot/Bootstrap/main.cpp]:
```suggestion
// so create a dummy one here and make it weak to allow wasi-libc to provide the real implementation for WASI reactor components.
```

**Review by jkotas — APPROVED** [review]
This looks good to me. @SingleAccretion ?

**Review by SingleAccretion — COMMENTED** [review]
The code changes LGTM, but I do have a question about wasmtime's behavior w.r.t. `_initialize`.

On the surface of things, https://github.com/bytecodealliance/wasmtime/pull/1565 claims to execute for it reactors. Why does it not work in this case?

**Comment by yowl** [comment]
Yes, a good question, I will try to get a more definitive answer and update here, but my current understanding is that wasmtime will call `_initialize` when it is invoking a WASI reactor module, but that is not the same thing as a WASM component (components are a lower level and dont require WASI).  To invoke the c# wasm component it is passed to the instantiation of another module, as in the example https://docs.rs/wasmtime/latest/wasmtime/component/macro.bindgen.html .

**Pull request merged**